File: Stopwatch/app/src/main/java/com/zukron/stopwatch/activity/MainActivity.java
package com.zukron.stopwatch.activity;

import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ListView;

import com.triggertrap.seekarc.SeekArc;
import com.zukron.stopwatch.adapter.LapAdapter;
import com.zukron.stopwatch.R;
import com.zukron.stopwatch.model.Lap;

import java.util.ArrayList;
import java.util.Locale;

public class MainActivity extends AppCompatActivity implements View.OnClickListener {
    private SeekArc saTime;
    private Button btnTime, btnPause, btnLapReset;
    private ListView lvLapResult;
    private static boolean isStart = false;
    private static int increment = 0;
    private static int indexLap = 0;
    private ArrayList<Integer> incrementArray = new ArrayList<>();
    private ArrayList<Lap> laps = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        saTime = findViewById(R.id.sa_time);
        btnTime = findViewById(R.id.btn_time);
        btnPause = findViewById(R.id.btn_pause);
        btnLapReset = findViewById(R.id.btn_lap_reset);
        lvLapResult = findViewById(R.id.lv_lap_result);

        init();
    }

    private void init() {
        incrementArray.add(0);

        saTime.setProgress(0);
        saTime.setEnabled(false);

        btnPause.setEnabled(false);
        btnLapReset.setEnabled(false);

        btnTime.setOnClickListener(this);
        btnPause.setOnClickListener(this);
        btnLapReset.setOnClickListener(this);
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btn_time:
                startStopwatch();
                break;
            case R.id.btn_pause:
                pauseStopwatch();
                break;
            case R.id.btn_lap_reset:
                lapResetStopwatch();
                break;
        }
    }

    private void startStopwatchThread() {
        Thread stopwatchThread = new Thread() {
            @Override
            public void run() {
                while (MainActivity.isStart) {
                    try {
                        Thread.sleep(10);

                        runOnUiThread(new Runnable() {
                            @Override
                            public void run() {
                                int minute = convertToMinute(MainActivity.increment);
                                int second = convertToSecond(MainActivity.increment);
                                int milli = convertTOMilli(MainActivity.increment);

                                double progress = ((double) second / 60) * 100;
                                saTime.setProgress((int) progress);

                                btnTime.setText(formatToDuration(minute, second, milli));
                            }
                        });

                        MainActivity.increment += 1;
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            }
        };

        stopwatchThread.start();
    }

    private int convertToMinute(int i) {
        return (i / 90) / 60;
    }

    private int convertToSecond(int i) {
        return (i / 90) % 60;
    }

    private int convertTOMilli(int i) {
        return (i % 90);
    }

    private String formatToDuration(int minute, int second, int milli) {
        return String.format(Locale.US, "%02d:%02d:%02d", minute, second, milli);
    }

    private void pauseStopwatch() {
        MainActivity.isStart = false;

        btnLapReset.setText(getResources().getString(R.string.reset));
    }

    private void startStopwatch() {
        btnPause.setEnabled(true);
        btnLapReset.setEnabled(true);
        btnLapReset.setText(getResources().getString(R.string.lap));

        MainActivity.isStart = true;
        startStopwatchThread();
    }

    private void lapResetStopwatch() {
        if (btnLapReset.getText().equals(getResources().getString(R.string.lap))) {
            // increment array untuk current lap
            incrementArray.add(MainActivity.increment);

            int curInc = incrementArray.get(MainActivity.indexLap + 1);
            int prevInc = incrementArray.get(MainActivity.indexLap);

            Lap lap = new Lap();
            lap.setIndex(MainActivity.indexLap);
            lap.setLap(MainActivity.increment);
            lap.setDiff(curInc - prevInc);

            laps.add(lap);

            LapAdapter lapAdapter = new LapAdapter(this, laps);
            lvLapResult.setAdapter(lapAdapter);

            // increment index
            MainActivity.indexLap += 1;
        } else if (btnLapReset.getText().equals(getResources().getString(R.string.reset))) {
            MainActivity.increment = 0;
            MainActivity.indexLap = 0;

            saTime.setProgress(0);
            laps.clear();
            incrementArray.clear();
            incrementArray.add(0);

            lvLapResult.setAdapter(null);

            btnPause.setEnabled(false);
            btnLapReset.setEnabled(false);
            btnTime.setText(getResources().getString(R.string.start));
        }
    }
}

File: Stopwatch/app/src/main/java/com/zukron/stopwatch/activity/MainActivity.kt
package com.zukron.stopwatch.activity

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import com.zukron.stopwatch.R
import com.zukron.stopwatch.adapter.LapAdapter
import com.zukron.stopwatch.model.Lap
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.coroutines.*
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.Dispatchers.Main
import kotlin.properties.Delegates

/**
 * Project name is Stopwatch
 * Created by Zukron Alviandy R on 9/2/2020
 * Contact me if any issues on zukronalviandy@gmail.com
 */
class MainActivity : AppCompatActivity(), View.OnClickListener {
    private var running: Boolean = false
    private var increment = 0
    private var curIncrement = 0
    private var prevIncrement = 0
    private var incrementLap: Int by Delegates.observable(0) { _, oldValue, newValue ->
        curIncrement = newValue
        prevIncrement = oldValue
    }

    private var indexLap = 0
    private var listLap: MutableList<Lap> = mutableListOf()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
    }

    override fun onStart() {
        super.onStart()

        sa_time.progress = 0
        sa_time.isEnabled = false

        btn_pause.isEnabled = false
        btn_lap_reset.isEnabled = false

        btn_time.setOnClickListener(this)
        btn_pause.setOnClickListener(this)
        btn_lap_reset.setOnClickListener(this)
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.btn_time -> startStopwatch()
            R.id.btn_pause -> pauseStopwatch()
            R.id.btn_lap_reset -> lapOrResetStopWatch()
        }
    }

    private fun startStopwatch() {
        btn_pause.isEnabled = true
        btn_lap_reset.isEnabled = true
        btn_lap_reset.text = resources.getString(R.string.lap)

        // jika state kondisi false atau pause, untuk menghindari start yang berulang" jika user menekan tombol start
        if (!running) {
            running = true
            CoroutineScope(IO).launch {
                while (running) {
                    delay(10)

                    val second = (increment / 90) % 60
                    val progress = (second / 60) * 100

                    withContext(Main) {
                        sa_time.progress = progress
                        btn_time.text = Lap.convertToDuration(increment)
                    }

                    increment += 1
                }
            }
        }
    }

    private fun pauseStopwatch() {
        running = false
        btn_lap_reset.text = resources.getString(R.string.reset)
    }

    private fun lapOrResetStopWatch() {
        if (running) {
            incrementLap = increment
            val diff = curIncrement - prevIncrement

            val lap = Lap(indexLap, increment, diff)
            listLap.add(lap)

            val lapAdapter = LapAdapter(this, listLap)
            lv_lap_result.adapter = lapAdapter

            indexLap += 1
        } else {
            increment = 0
            incrementLap = 0
            indexLap = 0
            listLap.clear()

            sa_time.progress = 0
            lv_lap_result.adapter = null

            btn_pause.isEnabled = false
            btn_lap_reset.isEnabled = false
            btn_time.text = resources.getString(R.string.start)
        }
    }
}
